src/Workspaces/Remote/Core/Diagnostics/DiagnosticComputer.cs

```diff
@@ -124,28 +164,99 @@ private static ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(BidirectionalMap<
             return builder.ToImmutable();
         }
 
-        private static BidirectionalMap<string, DiagnosticAnalyzer> CreateAnalyzerMap(Project project)
+        private sealed class CompilationWithAnalyzersData
         {
-            // we could consider creating a service so that we don't do this repeatedly if this shows up as perf cost
-            using var pooledObject = SharedPools.Default<HashSet<object>>().GetPooledObject();
-            using var pooledMap = SharedPools.Default<Dictionary<string, DiagnosticAnalyzer>>().GetPooledObject();
-            var referenceSet = pooledObject.Object;
-            var analyzerMap = pooledMap.Object;
-
-            // this follow what we do in DiagnosticAnalyzerInfoCache.CheckAnalyzerReferenceIdentity
-            foreach (var reference in project.Solution.AnalyzerReferences.Concat(project.AnalyzerReferences))
+            /// <summary>
+            /// Cache of <see cref="CompilationWithAnalyzers"/> and a map from analyzer IDs to <see cref="DiagnosticAnalyzer"/>s
+            /// for all analyzers for each project.
+            /// The <see cref="CompilationWithAnalyzers"/> instance is shared between all the following analyses modes for the project:
+            ///  1. Span-based analysis for active document (lightbulb)
+            ///  2. Background analysis for active and open documents.
+            ///  3. Background project analysis.
+            ///  4. <see cref="DefaultDiagnosticAnalyzerService"/> computation for analyzer diagnostics.
+            /// </summary>
+            private static readonly ConditionalWeakTable<Project, CompilationWithAnalyzersData> s_cache
```

Comment:
This is the core shared cache for all our OOP diagnostic computation. All diagnostic computations required in the IDE, regardless of the scope or the client, will now go through this cache and the returned `CompilationWithAnalyzers` instance from here.